registry/storage/driver/s3-aws/s3.go

```diff
@@ -1338,6 +1338,7 @@ func (b *buffer) Clear() {
 // cleanly resumed in the future. This is violated if Close is called after less
 // than a full chunk is written.
 type writer struct {
+	ctx       context.Context
```

Comment:
https://go.dev/blog/context-and-structs

Reply:
Yes, please read the whole thread in the discussion linked in the PR description. It's also why this PR is marked as a Draft. @corhere and I wanted to chat some things through and code is the best place to discuss them 😄

Reply:
That doesn't really explain it for me. Why have the context in the object instead of adding a ctx param to every method of writer that makes an s3 call?

Reply:
Because `writer` is created in S3 driver dynamically in the `(driver).Writer()` call
* https://github.com/distribution/distribution/blob/d24243730fad0fa858da3e26dbb7b720ee72aa2a/registry/storage/driver/s3-aws/s3.go#L699
* https://github.com/distribution/distribution/blob/d24243730fad0fa858da3e26dbb7b720ee72aa2a/registry/storage/driver/s3-aws/s3.go#L746

Now, it ten calls a bunch of AWS SDK funcs (that fire API reqs to AWS) that need to have `context` passed down to them from the upstream consumer so they can be called etc,.. The biggest problem is the `Write` method that must implement `io.Writer`:
* https://github.com/distribution/distribution/blob/d24243730fad0fa858da3e26dbb7b720ee72aa2a/registry/storage/driver/s3-aws/s3.go#L1406

And then there is also `flush` which also needs the upstream context because it makes upload finalising API calls and is called from the above-mentioned `Write` and from `Close` (which is there to implement `io.Closer` IIRC).

How else do imagine this works ?

Reply:
I don't have a better solution, but I will need to read a little more code before I understand it enough to review confidently :)

Reply:
SGTM. I hope you don't get too horrified by all that code 😁 No rush on this one

Reply:
Had a look, apologies for just pointing at generic doc instead of reading properly the first time.

I guess this is similar to solutions like https://pkg.go.dev/github.com/dolmen-go/contextio

One thing that could be a nice addition is a test that starts a writer and then cancels it to make sure we get back a context cancellation error, but that's not blocking for me.